### ISSUE TITLE
Start fails with "Unable to resolve service for type 'ChannelServiceHandler' while attempting to activate 'ChannelServiceController'.)'

### ISSUE THREAD

**Issue by sibbl** [microsoft/botbuilder-dotnet#3256]
## Version
4.7.0

## Describe the bug
I'm writing a ASP.NET Core 3.1 C# app where I use several packages of the BotBuilder SDK. What works fine with SDK 4.6.3 now fails with 4.7.0 with the following exception:

```
System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController Lifetime: Transient ImplementationType: Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController': Unable to resolve service for type 'Microsoft.Bot.Builder.ChannelServiceHandler' while attempting to activate 'Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController'.)'

Exception 1/2

InvalidOperationException: Error while validating the service descriptor 'ServiceType: Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController Lifetime: Transient ImplementationType: Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController': Unable to resolve service for type 'Microsoft.Bot.Builder.ChannelServiceHandler' while attempting to activate 'Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController'.

Exception 2/2

InvalidOperationException: Unable to resolve service for type 'Microsoft.Bot.Builder.ChannelServiceHandler' while attempting to activate 'Microsoft.Bot.Builder.Integration.AspNet.Core.ChannelServiceController'.



This exception was originally thrown at this call stack:
	Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(System.Type, System.Type, Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteChain, System.Reflection.ParameterInfo[], bool)
	Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateConstructorCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.ResultCache, System.Type, System.Type, Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteChain)
	Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateExact(Microsoft.Extensions.DependencyInjection.ServiceDescriptor, System.Type, Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteChain, int)
	Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.GetCallSite(Microsoft.Extensions.DependencyInjection.ServiceDescriptor, Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteChain)
	Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.ValidateService(Microsoft.Extensions.DependencyInjection.ServiceDescriptor)


```

## Expected behavior
The application starts fine.

## Additional context
Is it possible that this is related to this PR? https://github.com/microsoft/botbuilder-dotnet/pull/3038/files

I also hope that I can build a project to reproduce the error. It currently also only fails when compiling a debug version and running it with the ASP Environment set to `Development` for whatever reason.

Is there anything I possible did wrong or a cache I need to clear? Like I said, the same code runs fine with version 4.6.3 of all Microsoft.Bot.* packages.

[bug]

**Comment by dmvtech** [comment]
hi @sibbl 

I would be interested in a repro as well. What botbuilder packages are you using? Are you starting from a visx project or a sample?

> It currently also only fails when compiling a debug version and running it with the ASP Environment set to Development for whatever reason

That is interesting. Is all this testing, etc done locally? (running bot locally and testing with emulator)

I don't believe this has anything to do with 3.1 or the linked issue, but we'll see. If we can reproduce it, that will help. (If you're able to throw a repro up on a repo, even better).

**Comment by sibbl** [comment]
@dmvtech thanks for your quick reply! Please find a sample project here: 
[BotBuilderDemoProject.zip](https://github.com/microsoft/botbuilder-dotnet/files/4064070/BotBuilderDemoProject.zip)


You can reproduce it either by 
1. opening the solution and running it in Visual Studio 
2. or by running the `reproduce.ps1` which sets the asp.net env to "dev" and uses `dotnet` for building and running it.

While fiddling together the project, I found out that `services.AddMvc().AddControllersAsServices();` causes the trouble. Changing the line to `services.AddMvc();` fixes it. Then the whole bot should behave just like the `EchoBot` from the samples repo.

Possibly there's a circular dependency due to some changes in the `ChannelServiceController`?

One more thing: there's the special fact that we're using a common library ("UniBot" in the sample) with reusable dialogs, services, and also a common BotStartup implementation. This project is then referenced and used by other projects ("ImmoBot" in this sample). I was not able to reproduce the issue using a single project.

**Comment by EricDahlvang** [comment]
Hi @sibbl 

With Bot Builder 4.7, we introduced a [ChannelServiceController](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ChannelServiceController.cs).  This is being discovered by dependency injection when calling `.AddControllersAsServices()`

As a workaround, it is possible to exclude the integration library from discovery with something like:

```cs
public virtual void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc().ConfigureApplicationPartManager(manager=> 
                {
                    var botAssemblyPart = manager.ApplicationParts.FirstOrDefault(a => a.Name == typeof(ChannelServiceController).Assembly.GetName().Name);
                    manager.ApplicationParts.Remove(botAssemblyPart);
                })
            .AddControllersAsServices();
            
 
            services.AddSingleton<IBotFrameworkHttpAdapter, BotFrameworkHttpAdapter>();
 
            services.AddTransient<IBot, GenericBot>();
        }
```

Or by using a custom `ControllerFeatureProvider` to filter out the `ChannelServiceController`

We will raise this issue with the team, and come up with a better long term solution.

**Comment by johnataylor** [comment]
@gabog PR fixed this issue for 4.8. To work with 4.7 you can use the work around from @EricDahlvang

**Comment by gabog** [comment]
Hi @sibbl, we accelerated the fix for this issue and we just released a fix, if you update your project to use 4.7.2 it should work now.

Thanks for the feedback and detailed repro project.
